README.md
# Prism

**Warning: This is alpha software. Feedback encouraged!**

Prism is a fast, modular realtime (WebSocket) server for Node.js. Use it to power dashboards, games, and other realtime apps.

All functionality (RPC, PubSub, model syncing, etc) is provided by [Realtime Services](https://github.com/socketstream/realtime-service) - a new way to write realtime functionality and share it on `npm` without any vendor lock-in. Connect to a Prism server from the browser (or remote Node process) using [prism-client](https://github.com/socketstream/prism-client). 

Note: Prism is **just** a standalone realtime server, nothing else.  
SocketStream 0.4 uses Prism to offer a fully integrated developer experience.


## Example

```js
var prism = require('prism');
var rtt = require('rtt-engineio')();

var server = prism({port: 3001, transport: rtt});

// Add a Realtime Service called "rpc"
server.service('rpc', require('rts-rpc')());

server.start();
```

[View more examples](examples)

## Features

* designed for speed and scalability (see benchmarks below)
* no core functionality - everything is provided by modular [Realtime Services](https://github.com/socketstream/realtime-service)
* use native WebSockets, Engine.IO, SockJS, or any other [Realtime Transport](https://github.com/socketstream/realtime-transport)
* the [client](https://github.com/socketstream/prism-client) runs in the browser or a remote Node process
* share sessions between Prism and [Express.js](http://expressjs.com) (thanks to Connect Session Store)
* swap to Connect Redis Store (or similar) when you want to scale out
* generates a custom client module containing all the code you need for the browser
* prevent DDOS attacks and drop malicious messages with Connect-style middleware
* super-efficient protocol to reduce bytes over the wire (doesn't force JSON)
* optionally sends errors back to the client for easy debugging
* shared event bus so your application code can react to client disconnects
* ultra-light client-side code (for sending to browser)


## API

[View API documentation](API.md) (generated from source code)


## Truly Modular

Prism uses [Realtime Services](https://github.com/socketstream/realtime-service), a new way to provide your app with PubSub, RPC, Realtime Model functionality and more without any vendor lock-in. Realtime Services can be easily written, tested and shared on `npm`.

Combine multiple Realtime Services together to create the exact functionality your app needs. Prism automatically handles the message multiplexing for you using minimal bytes and CPU.


## Connecting from the browser

Once you've added your transport and services, Prism can (optionally) build a custom client module for you:

```js
var client = server.buildClient({});
fs.writeFileSync('client.js', client, 'utf8');
```

This module contains all the client-side code you need, based upon your choice of transport and services (minus any libraries which sadly can't be wrapped as Common JS modules - see Browser Assets below).

Use the `client.js` module you've just built in your application's client-side code as so:

```js
var app = require('./client')();
app.connect(function(err, info) {
  console.log('Connected to the server!', info);
  console.log('You can now call this API in your client code:', app.api);
});
```

Take this file, [browserify](https://github.com/substack/node-browserify) it, [minimize](https://github.com/mishoo/UglifyJS2) it and host it on a CDN.

See the [examples](examples) directory or the `prism-client` Readme for more info.


## Sessions

Prism supports sessions using the Connect Session Store. This design choice was made to allow integration with Express.js and various authentication systems (such as [Passport](http://passportjs.org/) which use Connect.

By default we use the in-memory store. This is fine for development, but not in production. Switch to a scalable session store (e.g. Connect Redis) with:

```js
var app = new Server({
  sessionStore:   new RedisStore({port: 6379})
});
```

## Scaling

The recommended way to scale Prism is behind a proxy such as HAProxy or Nginx (which now supports WebSockets).

Prism requires the load balancer to support 'sticky sessions', ensuring subsequent connections from the same browser (identified by the `connect.sid` cookie) are routed to the same back-end server. This gives a huge performance boost as all session data can be cached in RAM (although any changes are always saved back to the Session Store).

Alternatively, you're able to route all incoming requests to any back-end server you please. However, under this setup you'll have to disable in-memory session caching (with `cacheSession: false`). This will force Prism to query the Session Store on each incoming message. This will naturally be much slower, unless the Session Store is designed to be used in a distributed environment.


## Request Middleware

As incoming requests hit the server, you are able to inspect them, drop them or transform them before they are sent on to the relevant Realtime Service for processing. For performance reasons this functionality is implemented as middleware, exactly like in Express.

Available middleware modules:

[**prism-rate-limiter**](https://github.com/socketstream/prism-rate-limiter) - limits requests per second to help protect against DOS attacks

Use middleware as so:

```js
server.use(require('prism-rate-limiter')());
```


## Browser Assets

Transports and Services often need to send assets to the browser. For example, a service designed to work with Backbone may want to send a particular version of `backbone.js` to the client.

To obtain a list of static assets to send, call:

```js
var assets = server.browserAssets();
```

Unless you're using SocketStream 0.4, you'll need manually transform this Array of Objects into something your build system can use. Take a look at how we've done that in the `/examples/with_express` folder.

It's worth pointing out this area of Prism may change before Version 1 is released. I'm currently evaluating Bower as a possible solution to this problem. Feedback welcome!


## Logging

Prism is silent by default. To turn on logging of each incoming request, pass `console.log`, or your own custom logging function, to the server on startup:

```js
var app = new Server({
  port:           3001,
  transport:      engineio,
  log:            console.log
});
```

## FAQs

### Why not use Transform Streams instead of Connect-style middleware?

I tried this initially. Not only are they more complex to implement, in practice four transform streams are almost 10 times slower than four pieces of middleware on the stack. 


## Benchmarks

Install dev dependencies then run them with `npm run benchmarks'.

On a 2009 iMac, output is currently as follows:


    Benchmark 1: Simple echo service
    ✓ Requests per second: 124000

    Benchmark 2: Simple echo service (with JSON)
    ✓ Requests per second: 61000

    Benchmark 3: Simple echo service (with callbacks)
    ✓ Requests per second: 69000

    Benchmark 4: Simple echo service (with sessions from cache)
    ✓ Requests per second: 75000


Note: While the numbers are not bad, I think we can do better.
If it is necessary to change the internal APIs to improve performance I
will do so before Version 1 is released. Contributions / thoughts appreciated.


## License 

MIT


HISTORY.md
0.0.1 / 2013-04-21
==================

First release

index.js
"use strict";

/*!
 * Prism Realtime Server
 *
 * Copyright(c) 2013 Owen Barnes <owen@socketstream.org>
 * MIT Licensed
 */


var fs = require('fs');
var path = require('path');
var EventEmitter = require('events').EventEmitter;
var Service = require('realtime-service');
var Request = require('./lib/request');
var systemService = require('./lib/system-service');


/**
 * Prism Server
 *
 * Example:
 *
 *     var server = new Server({
 *       root:          '/my/app/dir',
 *       dir:           'services',
 *       log:           console.log,
 *       transport:     require('rtt-engineio')(),
 *       events:        instanceOfAnEventEmitter,
 *       sessionStore:  new RedisStore({port: 6379}),
 *       cacheSessions: true
 *     });
 * 
 * @param {Object} options
 * @return {Object} instance of ServiceManager
 * @api public
 *  
 */

function Server(options) {
  options = options || {};

  this.serviceCount = 0;
  this.services = {};
  this.api = {};                                        // server-side api returned from services
  this.middleware = [];                                 // request middleware (processed before msg is sent to service)

  this.version = require('./package.json').version;
  this.root = options.root || process.cwd().replace(/\\/g, '/');
  this.transport = options.transport || null;
  this.port = options.port || 3001;
  this.config = options.config || {};                   // sent to the client
  this.events = options.events || new EventEmitter();

  this.dir = options.dir || 'services';                 // dir containing Service data files
  this.log = options.log || function(){};

  this.sessionStore = options.sessionStore || false;    // will use default Connect Memory store if not set
  this.cacheSessions = options.cacheSessions || true;   // cache sessions in RAM (avoids hitting the store on each req)
 
  this._registerSystemService();
}


/**
 * Use a Realtime Service
 *
 * Examples:
 *
 *     server.service('rpc', require('rts-rpc')())
 *
 * @param {String} name of service (must be unique and < 12 chars)
 * @param {Object} service definition object (see Realtime Service Spec)
 * @param {Object} options and overrides (e.g. don't send client libs). not implemeted yet
 * @return {Object} instance of Service
 * @api public
 *  
 */

Server.prototype.service = function(name, definition, options) {
  options = options || {};

  // check name length
  if (name.length > 12) throw new Error("Service name '" + name + "' must be 12 chars or less");

  // todo: check for spaces and other weird chars (as name must form a directory)

  // check for name uniqueness
  for (var serviceId in this.services) {
     if (this.services[serviceId].assigned.name === name)
       throw new Error('Service name ' + name + ' already used. Please choose another name.');
  }

  var assign = {
    id:           this.serviceCount++,
    name:         name,
    api:          this.api,
    events:       this.events,
    root:         options.root || path.join(this.root, this.dir, name),
    log:          options.log || this.log,
    options:      options || {},
    version:      this.version
  };

  var service = new Service(definition, assign, this);
  this.services[assign.id] = service;
  return service;
};


/**
 * Use pre-request middleware for rate limiting, message sanitizing etc
 *
 * Works exactly like Express/Connect middleware
 *
 * Examples:
 *
 *     server.use(require('ss-rate-limiter'));
 *
 * @param {Function} middleware to execute
 * @return {}
 * @api public
 *  
 */

Server.prototype.use = function(fn) {
  var middleware = fn(this);
  this.middleware.push(middleware);
};


/**
 * Returns a list of all files which need to be sent to the browser
 *
 * @return {Array}
 * @api public
 *  
 */

Server.prototype.browserAssets = function(options) {
  options = options || {};
  var output = this.transport.browserAssets || [];
  for (var id in this.services) {
    var service = this.services[id];
    if (service.browserAssets && service.browserAssets.length) {
      output = output.concat(service.browserAssets);
    }
  }
  return output;
};


/**
 * Builds a custom client module which includes the client-side
 * transport and service code you need
 *
 * @return {String}
 * @api public
 *  
 */

Server.prototype.buildClient = function() {
  return require('./lib/code-generator')(this);
};



/**
 * Returns an array of non-private services which should be sent to the client
 *
 * @return {Array}
 * @api public
 *  
 */

Server.prototype.publicServices = function() {
  var buf = [];
  for (var id in this.services) {
    var service = this.services[id];
    if (service.private) continue;
    buf.push(service);
  }
  return buf;
};



/**
 * Process Incoming Request from a Realtime Transport
 *
 * Examples:
 *
 *     server.process({message: '1|{"method": "callMe"}', socketId: 1234});
 *
 * @param {Object} an object containing the raw message plus info about the sender (socketId, transport, etc)
 * @return {null}
 * @api public
 *  
 */

Server.prototype.process = function(params) {
  var request = new Request(params, this);
  execute(this.middleware, request, function(msg){
    this.api._system.sendError(request.socketId, msg);
    this.log('error', msg, 'for socketId', request.socketId);
  }.bind(this));
};


/**
 * Start Server
 *
 * Examples:
 *
 *     server.start(function() {
 *       console.log("Realtime Server started on port %s", server.port);
 *     });
 *
 * @param {Function} Callback to execute once server has started
 * @return {Object} Server-side API
 * @api public
 *  
 */

Server.prototype.start = function(cb) {
  if (!this.transport) throw new Error('The server can only be started once a Realtime Transport has been defined. Set with server.transport()');

  var connection = this.transport.server(this);

  for (var id in this.services) {
    var service = this.services[id];
    var serverApi = service.start(connection);
    if (serverApi) this.api[service.assigned.name] = serverApi;
  }

  // Add main request processor
  this.middleware.push(function(req) {
    req.process();
  });

  if (cb) cb(); // todo: callback when server really starts

  // Return an API object of functions you can call in your server-side code
  return this.api;
};


Server.prototype._registerSystemService = function() {
  this.service('_system', systemService(this));
};


// Private Helpers
function execute(stack, request, cb) {
  function exec(req, res, i){
    stack[i].call(stack, req, res, function(){
      exec(req, res, ++i);
    });
  }
  exec(request, cb, 0);
}

module.exports = function(options){
  return new Server(options);
};